include/vcpkg/base/stringview.h

```diff
@@ -11,6 +11,10 @@
 
 namespace vcpkg
 {
+
+    // Forward declaration needed for StringView(T&) constructor
+    struct ZStringView;
```

Comment:
I don't think this is the correct fix since now there are multiple conversion paths between these. Can you post the actual compiler output and/or repro instructions for the error? Generally speaking there should not be overload sets that want both ZStringView and StringView.

Reply:
Using commit 90b8ebae, the following is generated when directly compiling vcpkg-tool.

```
-- The CXX compiler identification is GNU 8.3.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting the C++ compiler in use
-- Detecting the C++ compiler in use - gcc
-- Module support is disabled.
-- Version: 8.1.1
-- Build type: 
-- CXX_STANDARD: 17
-- Performing Test has_std_17_flag
-- Performing Test has_std_17_flag - Success
-- Performing Test has_std_1z_flag
-- Performing Test has_std_1z_flag - Success
-- Performing Test SUPPORTS_USER_DEFINED_LITERALS
-- Performing Test SUPPORTS_USER_DEFINED_LITERALS - Success
-- Performing Test FMT_HAS_VARIANT
-- Performing Test FMT_HAS_VARIANT - Success
-- Required features: cxx_variadic_templates
-- Looking for strtod_l
-- Looking for strtod_l - found
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /src/vcpkg-tool/build

cd build
make

Scanning dependencies of target fmt
[  0%] Building CXX object _deps/fmt-build/CMakeFiles/fmt.dir/src/format.cc.o
[  1%] Building CXX object _deps/fmt-build/CMakeFiles/fmt.dir/src/os.cc.o
[  2%] Linking CXX static library libfmt.a
[  2%] Built target fmt
Scanning dependencies of target vcpkglib
[  3%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/checks.cpp.o
[  0%] Building CXX object _deps/fmt-build/CMakeFiles/fmt.dir/src/format.cc.o
In file included from /src/vcpkg-tool/build/_deps/fmt-src/src/format.cc:8:
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format-inl.h: In function ‘int fmt::v8::detail::snprintf_float(T, int, fmt::v8::detail::float_specs, fmt::v8::detail::buffer<char>&)’:
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format-inl.h:2483:55: warning: useless cast to type ‘int’ [-Wuseless-cast]
       int fraction_size = static_cast<int>(end - p - 1);
                                                       ^
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format-inl.h:2512:64: warning: useless cast to type ‘int’ [-Wuseless-cast]
       fraction_size = static_cast<int>(fraction_end - begin - 1);
                                                                ^
In file included from /src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format-inl.h:29,
                 from /src/vcpkg-tool/build/_deps/fmt-src/src/format.cc:8:
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h: In instantiation of ‘Char* fmt::v8::detail::write_significand(Char*, UInt, int, int, Char) [with Char = char; UInt = long long unsigned int; typename std::enable_if<std::is_integral<UInt>::value, int>::type <anonymous> = 0]’:
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:1769:31:   required from ‘OutputIt fmt::v8::detail::write_significand(OutputIt, UInt, int, int, Char) [with OutputIt = fmt::v8::appender; UInt = long long unsigned int; Char = char; typename std::enable_if<(! std::is_pointer<typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> = 0]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:1848:29:   required from ‘OutputIt fmt::v8::detail::do_write_float(OutputIt, const DecimalFP&, const fmt::v8::basic_format_specs<Char>&, fmt::v8::detail::float_specs, fmt::v8::detail::locale_ref) [with OutputIt = fmt::v8::appender; DecimalFP = fmt::v8::detail::dragonbox::decimal_fp<double>; Char = char; Grouping = fmt::v8::detail::fallback_digit_grouping<char>]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:1933:57:   required from ‘OutputIt fmt::v8::detail::write_float(OutputIt, const DecimalFP&, const fmt::v8::basic_format_specs<Char>&, fmt::v8::detail::float_specs, fmt::v8::detail::locale_ref) [with OutputIt = fmt::v8::appender; DecimalFP = fmt::v8::detail::dragonbox::decimal_fp<double>; Char = char]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:2057:21:   required from ‘OutputIt fmt::v8::detail::write(OutputIt, T) [with Char = char; OutputIt = fmt::v8::appender; T = double; typename std::enable_if<fmt::v8::detail::is_fast_float<T>::value, int>::type <anonymous> = 0]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:2197:23:   required from ‘fmt::v8::detail::default_arg_formatter<Char>::iterator fmt::v8::detail::default_arg_formatter<Char>::operator()(T) [with T = double; Char = char; fmt::v8::detail::default_arg_formatter<Char>::iterator = fmt::v8::appender]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/core.h:1611:15:   required from ‘constexpr decltype (vis(0)) fmt::v8::visit_format_arg(Visitor&&, const fmt::v8::basic_format_arg<Context>&) [with Visitor = fmt::v8::detail::default_arg_formatter<char>; Context = fmt::v8::basic_format_context<fmt::v8::appender, char>; decltype (vis(0)) = fmt::v8::appender]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:2931:21:   required from ‘void fmt::v8::detail::vformat_to(fmt::v8::detail::buffer<T>&, fmt::v8::basic_string_view<Char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<typename std::conditional<std::is_same<typename fmt::v8::type_identity<T>::type, char>::value, fmt::v8::appender, std::back_insert_iterator<fmt::v8::detail::buffer<typename fmt::v8::type_identity<T>::type> > >::type, typename fmt::v8::type_identity<T>::type> >, fmt::v8::detail::locale_ref) [with Char = char; typename fmt::v8::type_identity<T>::type = char; typename std::conditional<std::is_same<typename fmt::v8::type_identity<T>::type, char>::value, fmt::v8::appender, std::back_insert_iterator<fmt::v8::detail::buffer<typename fmt::v8::type_identity<T>::type> > >::type = fmt::v8::appender]’
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format-inl.h:2588:39:   required from here
/src/vcpkg-tool/build/_deps/fmt-src/include/fmt/format.h:1750:36: warning: conversion from ‘long long unsigned int’ to ‘size_t’ {aka ‘unsigned int’} may change value [-Wconversion]
     copy2(out, digits2(significand % 100));
                        ~~~~~~~~~~~~^~~~~
[  1%] Building CXX object _deps/fmt-build/CMakeFiles/fmt.dir/src/os.cc.o
[  2%] Linking CXX static library libfmt.a
[  2%] Built target fmt
[  3%] Building CXX object CMakeFiles/vcpkglib.dir/src/vcpkg/base/checks.cpp.o
In file included from /src/vcpkg-tool/include/vcpkg/base/format.h:3,
                 from /src/vcpkg-tool/include/vcpkg/base/messages.h:7,
                 from /src/vcpkg-tool/include/vcpkg/base/basic_checks.h:4,
                 from /src/vcpkg-tool/include/vcpkg/base/checks.h:3,
                 from /src/vcpkg-tool/src/vcpkg/base/checks.cpp:1:
/src/vcpkg-tool/include/vcpkg/base/stringliteral.h: In member function ‘decltype (ctx.out()) fmt::v8::formatter<vcpkg::StringLiteral, Char, void>::format(const vcpkg::StringLiteral&, FormatContext&) const’:
/src/vcpkg-tool/include/vcpkg/base/fwd/format.h:19:82: error: call of overloaded ‘StringView(const vcpkg::StringLiteral&)’ is ambiguous
             return fmt::formatter<Base, Char, void>::format(static_cast<Base>(val), ctx);                              \
                                                                                  ^
/src/vcpkg-tool/include/vcpkg/base/stringliteral.h:20:1: note: in expansion of macro ‘VCPKG_FORMAT_AS’
 VCPKG_FORMAT_AS(vcpkg::StringLiteral, vcpkg::StringView);
 ^~~~~~~~~~~~~~~
In file included from /src/vcpkg-tool/include/vcpkg/base/format.h:7,
                 from /src/vcpkg-tool/include/vcpkg/base/messages.h:7,
                 from /src/vcpkg-tool/include/vcpkg/base/basic_checks.h:4,
                 from /src/vcpkg-tool/include/vcpkg/base/checks.h:3,
                 from /src/vcpkg-tool/src/vcpkg/base/checks.cpp:1:
/src/vcpkg-tool/include/vcpkg/base/stringview.h:17:9: note: candidate: ‘vcpkg::StringView::StringView(const string&)’
         StringView(const std::string& s) noexcept; // Implicit by design
         ^~~~~~~~~~
/src/vcpkg-tool/include/vcpkg/base/stringview.h:14:12: note: candidate: ‘constexpr vcpkg::StringView::StringView(const vcpkg::StringView&)’
     struct StringView
            ^~~~~~~~~~
/src/vcpkg-tool/include/vcpkg/base/stringview.h:14:12: note: candidate: ‘constexpr vcpkg::StringView::StringView(vcpkg::StringView&&)’
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-redundant-move’
make[2]: *** [CMakeFiles/vcpkglib.dir/build.make:63: CMakeFiles/vcpkglib.dir/src/vcpkg/base/checks.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:155: CMakeFiles/vcpkglib.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

Reply:
Hmmm OK so these are the conversion sequences gcc8 is considering:

* StringLiteral -> std::string -> print that
* StringLiteral -> ZStringView -> print that

The second one is clearly the better match because it's a derived to base conversion instead of a  user defined conversion to get there, but compiler bugs gonna compiler bug.

Ultimately I think the correct fix is that StringLiteral should not have an `operator std::string` to create the problem in the first place. However, when I make that change a lot of stuff is broken.... I still think fixing that lot of broken stuff is preferrable to creating the kind of circular reference described in this PR though. Let me ask around.

Reply:
What happens if you change `VCPKG_FORMAT_AS` to reference `ZStringView` rather than `StringView` in `stringliteral.h`?

Reply:
Changing the reference as requested does not work.

>error: redefinition of ‘struct fmt::v8::formatter<vcpkg::ZStringView, Char, void>’

This conflict occurs with the macro call in zstringview.h.

Reply:
I believe you changed:

```patch
-VCPKG_FORMAT_AS(vcpkg::StringLiteral, vcpkg::StringView);
+VCPKG_FORMAT_AS(vcpkg::ZStringView, vcpkg::StringView);
```

but what I believe Billy means is

```patch
-VCPKG_FORMAT_AS(vcpkg::StringLiteral, vcpkg::StringView);
+VCPKG_FORMAT_AS(vcpkg::StringLiteral, vcpkg::ZStringView);
```

Reply:
You're right, I did try the former last night.  With the latter, the original error is still seen.

Reply:
OK, in that case I think the only correct fix is to remove the implicit operator std::string from StringLiteral

(And WOW I can't believe we have an implicit conversion operator to std::string!)

Reply:
Can you try https://github.com/microsoft/vcpkg-tool/pull/448 and let me know if it fixes that part of the problem?

Reply:
@BillyONeal I tried #448 by itself. The issue continues to surface once checks.cpp is compiled.